digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectValidator.scala

```diff
@@ -577,27 +577,25 @@ class ProjectValidator {
 
     def checkMinMaxTrackRoadTypes(trackInterval: Seq[ProjectLink]): Option[ProjectLink] = {
       val (left, right) = trackInterval.partition(_.track == Track.LeftSide)
-
-      val leftRoadTypes = left.sortBy(_.startAddrMValue).map(_.roadType.value).foldLeft(Seq.empty[Int]) { case (list, next) =>
-        if (list.nonEmpty && list.last == next)
-          list
-        else
-          list :+ next
+      val sortedLeft = left.sortBy(_.startAddrMValue)
+      val sortedRight = right.sortBy(_.startAddrMValue)
+
+      val leftIter = sortedLeft.iterator
+      val rightIter = sortedRight.iterator
+      while (leftIter.hasNext && rightIter.hasNext) {
+        val leftLink = leftIter.next
+        val rightLink = rightIter.next
+        if (leftLink.roadType != rightLink.roadType)
+          return Some(leftLink)
       }
-
-      val rightRoadTypes = right.sortBy(_.startAddrMValue).map(_.roadType.value).foldLeft(Seq.empty[Int]) { case (list, next) =>
-        if (list.nonEmpty && list.last == next)
-          list
-        else
-          list :+ next
+      if (!leftIter.hasNext && !rightIter.hasNext) {
+        None
+      } else if (sortedLeft.size > sortedRight.size) {
```

Comment:
do we want to check sizes of the list? because in two tracks its normal to have more links in one side than in the other, which will create false positive error messages for this cases